### ISSUE TITLE
Can this library solve any convex-concave problem?

### ISSUE THREAD

**Issue by zhumingpassional** [cvxgrp/dccp#77]
Good work. 

Can this library solve any convex-concave problem? For example, there are non-convex constraints: 

(x + y)^2 -x^2 - y^2 - (x^2 + y^2) ^ (3/2) <= 0

-x^2 - y^2 + 100 <= 0

Is obtained solution globally optimal? 

Is there any paper to describe the algorithm/method? 

Thank you very much.

**Comment by stephenpboyd** [comment]
Yes, the paper is at https://web.stanford.edu/~boyd/papers/dccp.html.

> On Mar 24, 2022, at 2:45 AM, Ming Zhu ***@***.***> wrote:
> 
> 
> Good work.
> 
> Can this library solve any convex-concave problem? For example, there are non-convex constraints:
> 
> (x + y)^2 -x^2 - y^2 - (x^2 + y^2) ^ (3/2) <= 0
> 
> -x^2 - y^2 + 100 <= 0

Yes, of course.  Just read the documentation.
> 
> Is obtained solution globally optimal?

Ans no, of course not.  Again, please read the documentation and the paper.
> 
> Is there any paper to describe the algorithm/method?

See above.

**Comment by xinyueshen** [comment]
Hi @zhumingpassional, for the specific constraints that you mentioned, DCCP can handle them as long as you write them as
(x + y)^2 <= x^2 + y^2 + (x^2 + y^2) ^ (3/2)
and
100 <= x^2 + y^2.

**Comment by zhumingpassional** [comment]
@stephenpboyd @xinyueshen  

Thanks for your reply and valuable suggestions.  

Can CVXPYgen generate C code for embedded applications when using cvxpy and dccp?

Is the generated C code faster than the python code?

**Comment by xinyueshen** [comment]
Hi @zhumingpassional, thanks for the questions, but DCCP currently does not work with CVXPYgen.

**Comment by zhumingpassional** [comment]
Will CVXPYgen support DCCP in the near future? I really hope it can.

Thanks.

**Comment by zhumingpassional** [comment]
I use DCCP to solve a convex-concave problem, whose constraint is:
(x + y)^2 <= x^2 + y^2 + (x^2 + y^2) ^ (3/2)  # I used cvxpy.power(, 3/2)
It does not work. It raise errors:
" cvxpy.error.DCPError: Problem does not follow DCP rules. Specifically:
The following constraints are not DCP:
0.0 <= power(var4[0], 2.0) + power(var5[0], 2.0) , because the following subexpressions are not:
|--  0.0 <= power(var4[0], 2.0) + power(var5[0], 2.0)"

However, if I revise the constraint as 
(x + y)^2 <= x^2 + y^2 + (x^2 + y^2) ^ (2),  # I used cvxpy.power(, 2)
It works. 

Does this library only support the power function where the parameter should be integer? I am not sure.  

Hope to get your reply. Thanks very much.

**Comment by stephenpboyd** [comment]
Read the documentation, which is very clear about what expressions and constraints are accepted.  It is easy to handle yours, by simply rewriting it in the correct form.

**Comment by xinyueshen** [comment]
@zhumingpassional For the first question, the following code should work.
```
[(x + y) ** 2 <= x ** 2 + y ** 2 + cp.power(z, 3), z == cp.norm(cp.vstack([x, y]))]
```
For the second question, yes, you can initialize the variables by setting their values before calling DCCP to solve, and then DCCP will use the initialization that you provide. The code is simply like the following.
```
x.value = 1
y.value = 2
z.value = 3
prob.solve(method="dccp")
```

**Comment by zhumingpassional** [comment]
@stephenpboyd  @xinyueshen Thank you very much for your valuable reply. Excellent work! I will read the document carefully.
